### PR TITLE
Force load project were we don't have command line args in DPL

### LINKED ISSUE #19407: DeferredProjectWorkspaceService needs update to do the filtering on IVsSolutionWorkspaceService.GetManagedCommandLineInfoAsync()
IVsSolutionWorkspaceService.GetManagedCommandLineInfoAsync() used to return a dictionary with only valid intellisense data (CommandLineArgs and TargetPath not null or emtpy); the key is the full project path.

For 15.3 Preview 2, we've made the change to return the ManagedCommandLineInfo for all projects in the solution, so that when Roslyn calls this API, it can filter and act upon the projects that don't have valid Intellisense data, e.g. trigger a full load of those projects. Please make sure do the valid check to avoid crash.

### PR DESCRIPTION
**Pull request by Pilchie**
## Ask Mode
**Customer scenario**

In the case where we AnyCode is unable to determine the command line
arguments for a project (likely because something is different about the
way design time builds are set up), fall back to forcing the project to
load normally so that users still get a good experience from IDE
features.

**Bugs this fixes:**

Fixes #19407, and internal bug 432432.

**Workarounds, if any**

The customer can disable LSL, but we want people to be able to use it and continue to provide feedback.

**Risk**

This code should only be executed if LSL is enabled.

**Performance impact**

Low, because we still operate on the same set of projects, but we might skip some of them.

**Is this a regression from a previous update?**

Currently there is a crash, because AnyCode is now passing the failed projects to us.  

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

We didn't, it's planned work.

**How was the bug found?**

Planned work for 15.3 to address "broken" features with LSL enabled.

### CODE REVIEW

**Comment by Pilchie** [comment]
Tagging @dotnet/roslyn-ide for review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 1]
```diff
@@ -695,69 +695,72 @@ private void OutputToOutputWindow(string message)
...
+            // solution doesn't know about - be resilient to that case.
+            Guid projectGuid;
+            try
             {
-                return null;
+                projectGuid = solution5.GetGuidOfProjectFile(projectFilename);
+            }
+            catch (ArgumentException)
+            {
+                var message = $"Failed to get the project guid for '{projectFilename}' from the solution, using  random guid instead.";
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: .... What is the implication of this.  

How would we be unable to get a project guid?  Isn't it embedded in a project file?

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Won't this break anything in VS that needs these guids?  I don't understand the fallout of this.

Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Note that this code is just moved - The problem is that anycode might tell us about project (from it's index, not from the project file), that doesn't exist anymore.

Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Is that not an anycode bug?  Why don't they check and see if their index is valid before telling us about it?

**Comment by Pilchie** [comment]
Also tagging @huizhonglong for a review.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 1 above)
Comment (huizhonglong) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: AnyCode checked and passed only valid commandLineInfo before; now we pass all so that Roslyn can reload those projects that don't have valid commandLineInfo.

**Comment by huizhonglong** [comment]
I would suggest use EnsureProjectsAreLoaded() to load projects once to have some perf gain over EnsureProjectIsLoaded(). Other than that looks good to me.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 2]
```diff
@@ -862,6 +890,27 @@ private void OutputToOutputWindow(string message)
...
+            // If the index is stale, it might give us a path that doesn't exist anymore that the
+            // solution doesn't know about - be resilient to that case.
+            Guid projectGuid;
+            try
+            {
+                projectGuid = solution5.GetGuidOfProjectFile(projectFilename);
+            }
+            catch (ArgumentException)
+            {
+                var message = $"Failed to get the project guid for '{projectFilename}' from the solution, using  random guid instead.";
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Can we log the GUID so we can correlate dumps to the output window if we're trying to debug this?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 3]
```diff
@@ -626,26 +623,68 @@ private void InitializeOutputPane()
             cancellationToken.ThrowIfCancellationRequested();
             OutputToOutputWindow($"Getting project information - done (took {DateTimeOffset.UtcNow - start})");
 
+            ForceLoadProjectsWhoseDesignTimeBuildFailed(projectInfos);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Does this force load dependent projects? If B depends on A, and B failed for some reason, will this trigger a load on A which _didn't_ fail but CreateDeferredProjects might do something bad with now?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 4]
```diff
@@ -626,26 +623,68 @@ private void InitializeOutputPane()
...
+
+            CreateDeferredProjects(projectInfos, cancellationToken);
+
+            CallWithTimingLog("Pushing to workspace", FinishLoad);
+        }
+
+        private void ForceLoadProjectsWhoseDesignTimeBuildFailed(IReadOnlyDictionary<string, DeferredProjectInformation> projectInfos)
+        {
+            OutputToOutputWindow($"Loading projects whose design time builds failed - start");
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Not necessary for this PR but right now we have so many start/end/logging stuff here it's getting hard to follow. A disposable block that does timing might be nice.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs [hunk 5]
```diff
@@ -626,26 +623,68 @@ private void InitializeOutputPane()
...
+            var start = DateTimeOffset.UtcNow;
+            var solution4 = (IVsSolution4)_vsSolution;
+            var solution7 = (IVsSolution7)_vsSolution;
+            var guidsOfProjectsThatFailed = projectInfos.Where(pi => DesignTimeBuildFailed(pi.Value) && !solution7.IsDeferredProjectLoadAllowed(pi.Key)).Select(pi => GetProjectGuid(pi.Key)).ToArray();
+            OutputToOutputWindow($"\tForcing load of {guidsOfProjectsThatFailed.Length} projects.");
+            OutputListToOutputWindow("\tIncluding ", projectInfos.Where(pi => DesignTimeBuildFailed(pi.Value) && !solution7.IsDeferredProjectLoadAllowed(pi.Key)).Select(pi => pi.Key));
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Duplicate logic as line 639; extract it out.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 2 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Note that the output window stuff is `[Conditional("DEBUG")]` so it won't be there in dumps.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 5 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: I deleted your comment!?!  I'll extract the local.  Didn't before because of the `Conditional("DEBUG")`.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 4 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Will try that out and see how it looks

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 3 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: Not without passing `VSBSLFLAGS_LoadBuildDependencies` to `IVsSolution4.EnsureProjectsAreLoaded`.

Also `CreateDeferredProjects` should handle that fine - if the project is "real" loaded, it will no-op.  This is also necessary for:

1. Projects that don't support deferred loading.
2. Projects that the user might have triggered loading before `CreateDeferredProjects` is called (either explicitly by expanding them, or because they were expanded in the .suo file).

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs (hunk 4 above)
Comment (Pilchie) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectTracker.cs]: I don't like it.  #Won'tFix

**Comment by Pilchie** [comment]
Tagging @MattGertz - this is the full, planned fix for internal bug 432432

**Comment by Pilchie** [comment]
Hit https://github.com/dotnet/roslyn/issues/18295 for VB.

**Comment by Pilchie** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by Pilchie** [comment]
retest windows_release_vs-integration_prtest  please

**Comment by MattGertz** [comment]
Adding the Director label, though this is already on JoC's approval list -- we'll just check to make sure he's copacetic with the validation _et al._
